### PR TITLE
Add masking

### PR DESCRIPTION
**Pull request by nvaytet**
This adds the framework for masking according to various coordinates (`pixel_id`, `tof`, `wavelength`, `two_theta`).

It also attempts to standardize the Powgen and dream-geant4 use cases, in terms of loading files and keeping logical dimensions.

This is part of #50, but it only implements the reduction workflow side of things, and not an interactive masking tool, which is defered to a later PR.

(Apologies about the noise with quotation marks, I accidentally ran `black` at some point on the files, after the config for `black` was removed due to the switch to `ruff`, and it therefore did not ignore quotation marks)

### CODE REVIEW

File: docs/user-guide/dream/dream-data-reduction.ipynb [hunk 1]
```diff
@@ -23,40 +23,63 @@
...
     "    # The upper bounds mode is not yet implemented.\n",
     "    UncertaintyBroadcastMode: UncertaintyBroadcastMode.drop,\n",
+    "    # Edges for binning in d-spacing\n",
+    "    DspacingBins: sc.linspace(\"dspacing\", 0.0, 2.3434, 201, unit=\"angstrom\"),\n",
+    "    # Mask in time-of-flight\n",
+    "    TofMask: lambda x: (x < sc.scalar(0.0, unit=\"ns\"))\n",
+    "    | (x > sc.scalar(86e6, unit=\"ns\")),\n",
+    "    # Mask in wavelength\n",
+    "    WavelengthMask: lambda x: (x > sc.scalar(2.0, unit=\"angstrom\"))\n",
+    "    & (x < sc.scalar(2.7, unit=\"angstrom\")),\n",
```
[the remark this answers is not in the document]
Comment (nvaytet) [reply docs/user-guide/dream/dream-data-reduction.ipynb:58]: No there is no artifact. I was just to illustrate things.
I probably should make a separate example, either in a new section at the bottom or in a new notebook.

**Comment by nvaytet** [comment]
@jl-wynen for now, the notebooks only work when computing d-spacing from positions, and not from a calibration file.
I looked at computing the other scattering coordinates from the d-spacing, if calibration is present, but it seemed more difficult than I thought.
Maybe you know of a good approach for this that I missed?

**Comment by jl-wynen** [comment]
I know an approach but it may not be good. For CIF files we need to write data in tof. So the idea was to do the reduction in d-spacing and then convert back to tof using fake positions. You could use a similar approach here where you essentially determine calibrated positions by converting back from d-spacing. 

There is some ambiguity here about how you define positions and tof, though. So you / we would have to make a judgement call. But since we need this anyway for CIF, we can use the same mechanism here.

File: src/ess/dream/io/nexus.py [hunk 2]
```diff
@@ -1,135 +1,156 @@
...
+    RawSource,
+    ReducibleDetectorData,
+    RunType,
+    SamplePosition,
+    SourcePosition,
+)
+from ess.reduce import nexus
+
+DETECTOR_BANK_RESHAPING = {
+    "endcap_backward_detector": lambda x: x.fold(
```
Comment (nvaytet) [inline src/ess/dream/io/nexus.py:32]: Make this into sizes, not lambdas

File: src/ess/powder/external/powgen/beamline.py [hunk 3]
```diff
@@ -6,9 +6,16 @@
...
-from ...types import CalibrationData, DetectorDimensions, RawCalibrationData
+from ...types import (
+    CalibrationData,
+    NeXusDetectorDimensions,
+    NeXusDetectorName,
+    RawCalibrationData,
+)
 from .types import DetectorInfo
 
+DETECTOR_BANK_SHAPES = {"powgen_detector": {"bank": 23, "column": 154, "row": 7}}
```
Comment (nvaytet) [inline src/ess/powder/external/powgen/beamline.py:17]: shapes -> sizes

File: src/ess/powder/grouping.py [hunk 4]
```diff
@@ -2,67 +2,33 @@
...
     TwoThetaBins,
 )
 
 
-def group_by_two_theta(
-    data: DspacingData[RunType], edges: TwoThetaBins
+def focus_data(
+    data: MaskedData[RunType],
+    dspacing_bins: DspacingBins,
+    twotheta_bins: Optional[TwoThetaBins] = None,
```
Comment (nvaytet) [inline src/ess/powder/grouping.py:21]: Maybe request a new data type to get data binned in theta, as in sans Qxy?

File: docs/user-guide/sns-instruments/POWGEN_data_reduction.ipynb [hunk 5]
```diff
@@ -50,19 +50,21 @@
...
-    "    ValidTofRange: sc.array(dims=['tof'], values=[0.0, 16666.67], unit='us'),\n",
-    "\n",
     "    # Edges for binning in d-spacing\n",
-    "    DspacingBins: sc.linspace('dspacing', 0.0, 2.3434, 200, unit='angstrom'),\n",
+    "    DspacingBins: sc.linspace(\"dspacing\", 0.0, 2.3434, 201, unit=\"angstrom\"),\n",
+    "    # Mask in time-of-flight\n",
+    "    TofMask: lambda x: (x < sc.scalar(0.0, unit=\"us\"))\n",
+    "    | (x > sc.scalar(16666.67, unit=\"us\")),\n",
+    "    # Mask in wavelength\n",
+    "    WavelengthMask: lambda x: (x > sc.scalar(0.18, unit=\"angstrom\"))\n",
```
Comment (nvaytet) [inline docs/user-guide/sns-instruments/POWGEN_data_reduction.ipynb:66]: remove wavelength mask

File: src/ess/dream/data.py [hunk 6]
```diff
@@ -2,38 +2,38 @@
 # Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
 """Data for tests and documentation with DREAM."""
 
-from pathlib import Path
+# from pathlib import Path
```
Comment (nvaytet) [inline src/ess/dream/data.py:5]: commented code

**Review by jl-wynen — APPROVED** [review]

**Pull request merged**